cmd/terramate/cli/telemetry/telemetry.go

```diff
@@ -0,0 +1,217 @@
+	AuthOIDCGithub
+	AuthOIDCGitlab
+	AuthAPIKey
+)
+
+type Message struct {
+	Platform PlatformType `json:"pla,omitempty"`
+	Auth     AuthType     `json:"aut,omitempty"`
+
+	// Signature is an anonymized unique id.
+	Signature string `json:"sig,omitempty"`
+
+	Arch string `json:"arc,omitempty"`
+	OS   string `json:"os,omitempty"`
+
+	// Command stores the invoked command.
+	Command string `json:"cmd,omitempty"`
+
+	// Details stores features or flags used by the command.
+	Details []string `json:"det,omitempty"`
+}
+
+func DetectPlatformFromEnv() PlatformType {
+	if isEnvVarSet("GITHUB_ACTIONS") {
+		return PlatformGithub
+	} else if isEnvVarSet("GITLAB_CI") {
+		return PlatformGitlab
+	} else if isEnvVarSet("CI") {
+		return PlatformGenericCI
+	} else {
+		return PlatformLocal
+	}
+}
+
+func DetectAuthTypeFromEnv(credpath string) AuthType {
+	if isEnvVarSet("ACTIONS_ID_TOKEN_REQUEST_TOKEN") {
+		return AuthOIDCGithub
+	} else if isEnvVarSet("TM_GITLAB_ID_TOKEN") {
+		return AuthOIDCGitlab
+	} else {
+		return getAuthProviderFromCredentials(credpath)
+	}
+}
+
+func ReadSignature(p string) string {
+	sigBytes, err := os.ReadFile(p)
+	if err == nil {
+		lines := strings.SplitN(string(sigBytes), "\n", 2)
+		if len(lines) > 0 {
+			return strings.TrimSpace(lines[0])
+		}
+	}
+	return ""
+}
+
+func GenerateOrReadSignature(cpsigfile, anasigfile string) (string, bool) {
+	logger := log.With().
+		Str("action", "GenerateOrReadSignature()").
+		Logger()
+
+	// Try reading existing signature.
+	anasig := ReadSignature(anasigfile)
+	if anasig != "" {
+		return anasig, false
+	}
+
+	// Create a new one. Bootstrap from checkpoint signature, otherwise create a new one.
+	var newsig string
+	cpsig := ReadSignature(cpsigfile)
+	if cpsig != "" {
+		newsig = cpsig
+	} else {
+		newsig = GenerateSignature()
+	}
+
+	if err := os.MkdirAll(filepath.Dir(anasigfile), 0755); err != nil {
+		logger.Debug().Err(err).Msg("failed to create directory for signature file")
+	}
+	if err := os.WriteFile(anasigfile, []byte(newsig+"\n\n"+userMessage+"\n"), 0644); err != nil {
+		logger.Debug().Err(err).Msg("failed to save signature file")
+	}
+
+	return newsig, true
+}
+
+func GenerateSignature() string {
+	buf := make([]byte, 16)
+	_, _ = rand.Read(buf)
+	return hex.EncodeToString(buf)
+}
+
+func isEnvVarSet(key string) bool {
+	val := os.Getenv(key)
+	return val != "" && val != "0" && val != "false"
+}
+
+func getAuthProviderFromCredentials(credfile string) AuthType {
+	_, err := os.Lstat(credfile)
+	if err != nil {
+		return AuthNone
+	}
+	contents, err := os.ReadFile(credfile)
+	if err != nil {
+		return AuthNone
+	}
+
+	var providerProbe struct {
+		Provider string `json:"provider"`
+	}
+	err = json.Unmarshal(contents, &providerProbe)
+	if err != nil {
+		return AuthNone
+	}
+
+	switch providerProbe.Provider {
+	case "Google":
+		return AuthIDPGoogle
+	case "GitHub":
+		return AuthIDPGithub
+	default:
+		// There could be room here for AuthInvalid or AuthUnknown here.
```

Comment:
```suggestion
		// There could be room here for AuthInvalid or AuthUnknown.
```

Reply:
Done in [859a1d7](https://github.com/terramate-io/terramate/pull/1981/commits/859a1d764f184e3dea5e94594cc0adce1c25ef87)